### PR TITLE
NULL should be variant type instead of always assuming STRING

### PR DESCRIPTION
**Pull request by marregui**
For issue: https://github.com/questdb/questdb/issues/973

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 87 / 89 (97.75%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCompiler.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/cairo/NullColumn.java|14|15|93.33%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToShortFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InSymbolFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/AbstractEntityCastFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMapRecord.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToFloatFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToByteFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDoubleFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToLong256FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/bool/InStrFunctionFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToIntFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToDateFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|40|40|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReaderSelectedColumnRecord.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/NullFunctionFactory.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToCharFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReaderRecord.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToStrFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/ContiguousVirtualMemory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/Function.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastBooleanToLongFunctionFactory.java|1|1|100.00%|

File: core/src/test/java/io/questdb/griffin/SqlCompilerTest.java [hunk 1]
```diff
@@ -2546,6 +2546,40 @@ public void testInsertAsSelectColumnSubset2() throws Exception {
...
+            testInsertAsSelect("nothing",
+                    // in this context null is the default type, STRING
+                    "create table x (a CHAR, b LONG)",
+                    "insert into x " +
+                            "select" +
+                            " null," +
+                            " null" +
+                            " from long_sequence(3)",
+                    "x");
+            Assert.fail();
```
Comment (ideoma) [inline core/src/test/java/io/questdb/griffin/SqlCompilerTest.java:2576]: This should not fail, null should not be treated as a STRING but as a LONG here, ideally... Or put a comment that it's not done yet

File: core/src/test/java/io/questdb/griffin/InsertNullTest.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                setUp();
+            }
+            try {
+                final String[] type = types[i];
+                assertQuery(
+                        "value\n",
+                        String.format("x where value != null", type[1]),
+                        String.format("create table x (value %s)", type[0]),
+                        null,
+                        String.format("insert into x select cast(null as %s) from long_sequence(%d)", type[0], NULL_INSERTS),
```
Comment (ideoma) [inline core/src/test/java/io/questdb/griffin/InsertNullTest.java:103]: We don't want the user to cast null to the column type, again ideally...

**Review by ideoma — CHANGES_REQUESTED** [review]
This is intersting and good approach. I'd request you search for `cast(null as ` in tests and add similar test without `cast`, there are quite few of those already

**Comment by marregui** [comment]
Hi @ideoma thank you for the review!
I am trying a different approach in a parallel branch, not pushed, where I explore adding a new `ColumnType` for NULL.
Technically, null is not a type, rather the absence of value for the supported types. However I would like to explore that avenue as well, for comparison. I will amend this PR also, to incorporate the feedback.

Thank you again!

**Comment by Alexis1956** [comment]
I didn't do anything wrong but you trying to kill me and set me up is wrong. Stop it now or I call the law on you all. Game over. And give me back my money. Or go to jail. And I didn't break anything but you did. I'm done being nice. And you didn't have my permission to sue ny in my name.

**Comment by bluestreak01** [comment]
the other PR is better!

**Pull request closed without merging**